// astro/dist/core/endpoint/index.js
import { renderEndpoint } from "../../runtime/server/index.js";
import { ASTRO_VERSION } from "../constants.js";
import { AstroCookies, attachToResponse } from "../cookies/index.js";
import { getParamsAndProps, GetParamsAndPropsError } from "../render/core.js";
const clientAddressSymbol = Symbol.for("astro.clientAddress");
function createAPIContext({
  request,
  params,
  site,
  props,
  adapterName
}) {
  return {
    cookies: new AstroCookies(request),
    request,
    params,
    site: site ? new URL(site) : void 0,
    generator: `Astro v${ASTRO_VERSION}`,
    props,
    redirect(path, status) {
      return new Response(null, {
        status: status || 302,
        headers: {
          Location: path
        }
      });
    },
    url: new URL(request.url),
    get clientAddress() {
      if (!(clientAddressSymbol in request)) {
        if (adapterName) {
          throw new Error(
            `clientAddress is not available in the ${adapterName} adapter. File an issue with the adapter to add support.`
          );
        } else {
          throw new Error(
            `clientAddress is not available in your environment. Ensure that you are using an SSR adapter that supports this feature.`
          );
        }
      }
      return Reflect.get(request, clientAddressSymbol);
    }
  };
}
async function call(mod, env, ctx) {
  const paramsAndPropsResp = await getParamsAndProps({
    mod,
    route: ctx.route,
    routeCache: env.routeCache,
    pathname: ctx.pathname,
    logging: env.logging,
    ssr: env.ssr
  });
  if (paramsAndPropsResp === GetParamsAndPropsError.NoMatchingStaticPath) {
    throw new Error(
      `[getStaticPath] route pattern matched, but no matching static path found. (${ctx.pathname})`
    );
  }
  const [params, props] = paramsAndPropsResp;
  const context = createAPIContext({
    request: ctx.request,
    params,
    props,
    site: env.site,
    adapterName: env.adapterName
  });
  const response = await renderEndpoint(mod, context, env.ssr);
  if (response instanceof Response) {
    attachToResponse(response, context.cookies);
    return {
      type: "response",
      response
    };
  }
  return {
    type: "simple",
    body: response.body,
    encoding: response.encoding,
    cookies: context.cookies
  };
}
function isRedirect(statusCode) {
  return statusCode >= 300 && statusCode < 400;
}
function throwIfRedirectNotAllowed(response, config) {
  if (config.output !== "server" && isRedirect(response.status)) {
    throw new Error(
      `Redirects are only available when using output: 'server'. Update your Astro config if you need SSR features.`
    );
  }
}
export {
  call,
  throwIfRedirectNotAllowed
};
