// astro/dist/core/errors/codes.js
var AstroErrorCodes = /* @__PURE__ */ ((AstroErrorCodes2) => {
  AstroErrorCodes2[AstroErrorCodes2["StaticRedirectNotAllowed"] = 2005] = "StaticRedirectNotAllowed";
  AstroErrorCodes2[AstroErrorCodes2["UnavailableInSSR"] = 2006] = "UnavailableInSSR";
  AstroErrorCodes2[AstroErrorCodes2["GenericRuntimeError"] = 3e3] = "GenericRuntimeError";
  AstroErrorCodes2[AstroErrorCodes2["CssSyntaxError"] = 4e3] = "CssSyntaxError";
  AstroErrorCodes2[AstroErrorCodes2["CssUnknownError"] = 4001] = "CssUnknownError";
  AstroErrorCodes2[AstroErrorCodes2["FailedToLoadModuleSSR"] = 5e3] = "FailedToLoadModuleSSR";
  AstroErrorCodes2[AstroErrorCodes2["ConfigError"] = 6e3] = "ConfigError";
  AstroErrorCodes2[AstroErrorCodes2["GenericMarkdownError"] = 7e3] = "GenericMarkdownError";
  AstroErrorCodes2[AstroErrorCodes2["MarkdownFrontmatterParseError"] = 7001] = "MarkdownFrontmatterParseError";
  AstroErrorCodes2[AstroErrorCodes2["UnknownCompilerError"] = 9e3] = "UnknownCompilerError";
  AstroErrorCodes2[AstroErrorCodes2["UnknownCompilerCSSError"] = 9001] = "UnknownCompilerCSSError";
  AstroErrorCodes2[AstroErrorCodes2["UnknownViteSSRError"] = 9002] = "UnknownViteSSRError";
  AstroErrorCodes2[AstroErrorCodes2["UnknownError"] = 9999] = "UnknownError";
  return AstroErrorCodes2;
})(AstroErrorCodes || {});
export {
  AstroErrorCodes
};
