// astro/dist/vite-plugin-load-fallback/index.js
import nodeFs from "fs";
import npath from "path";
function loadFallbackPlugin({
  fs,
  settings
}) {
  if (!fs || fs === nodeFs) {
    return false;
  }
  const tryLoadModule = async (id) => {
    try {
      return await fs.promises.readFile(cleanUrl(id), "utf-8");
    } catch (e) {
      try {
        return await fs.promises.readFile(id, "utf-8");
      } catch (e2) {
        try {
          const fullpath = new URL("." + id, settings.config.root);
          return await fs.promises.readFile(fullpath, "utf-8");
        } catch (e3) {
        }
      }
    }
  };
  return [
    {
      name: "astro:load-fallback",
      enforce: "post",
      resolveId(id, parent) {
        if (id.startsWith(".") && parent && fs.existsSync(parent)) {
          return npath.posix.join(npath.posix.dirname(parent), id);
        }
      },
      async load(id) {
        const source = await tryLoadModule(id);
        return source;
      }
    },
    {
      name: "astro:load-fallback-hmr",
      enforce: "pre",
      handleHotUpdate(context) {
        const read = context.read;
        context.read = async () => {
          const source = await tryLoadModule(context.file);
          if (source)
            return source;
          return read.call(context);
        };
      }
    }
  ];
}
const queryRE = /\?.*$/s;
const hashRE = /#.*$/s;
const cleanUrl = (url) => url.replace(hashRE, "").replace(queryRE, "");
export {
  loadFallbackPlugin as default
};
